keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,845 @@
+    // A volume that is both attached and mounted on a node is considered to be used o
+    // The same rule applies for a unique volume that is shared among multiple pods on
+    // If this field is not specified, then the supported number of volumes on this no
+    // +optional
+    Count *int32
+}
+```
+
+#### CSIDriver
+
+A new field, `NodeAllocatableUpdatePeriodSeconds`, will be added to the `CSIDriverSpec` struct. This field allows a CSI driver to specify the interval at which the Kubelet should periodically query a driver's `NodeGetInfo` RPC endpoint to update the `CSINode` object. If this field is not set, updates will only occur in response to volume attachment failures as a result of no capacity.
+
+```golang
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ...
+    // NodeAllocatableUpdatePeriodSeconds specifies the interval between periodic updates of
+    // the CSINode allocatable capacity for this driver. If not set, periodic updates
+    // are disabled, and updates occur only upon detecting capacity-related failures.
+    // The minimum allowed value for this field is 10 seconds.
+    // +optional
+    NodeAllocatableUpdatePeriodSeconds *metav1.Duration
+}
+```
+
+#### Validation Changes
+
+The [ValidateCSINodeUpdate](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/storage/validation/validation.go#L304) function in the API validation code path will be modified to allow updates to the `Allocatable.Count`
+field when the feature gate is enabled:
+
+```golang
+func ValidateCSINodeUpdate(new, old *storage.CSINode) field.ErrorList {
+    allErrs := ValidateCSINode(new)
+    
+    if utilfeature.DefaultFeatureGate.Enabled(features.MutableCSINodeAllocatableCount) {
+        for _, oldDriver := range old.Spec.Drivers {
+           for _, newDriver := range new.Spec.Drivers {
+                // Allow Allocatable.Count to be modified
+                // Ensure all other fields are unchanged
+            }
+        }
+    } else {
+        // Existing validation logic for when feature gate is disabled
+    }
+    return allErrs
+}
+```
+
+This updated logic allows the `Allocatable.Count` field to be modified when the feature gate is enabled, while ensuring all other fields remain immutable. When the feature gate is disabled, it falls back to the existing validation logic for backward compatibility.
+
+#### Volume Plugin Manager
+
+A new goroutine will be started in VolumePluginMgr’s [Run()](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L953) func if the `NodeAllocatableUpdatePeriodSeconds` is set to a nonzero value. This goroutine will periodically trigger updates to the `CSINode` object based on the specified interval:
+
+```golang
+func (pm *VolumePluginMgr) Run(stopCh <-chan struct{}) {
+    if pm.csiNodeUpdateInterval > 0 {
+        go wait.Until(pm.updateCSINodeInfo, pm.csiNodeUpdateInterval, stopCh)
```

Comment:
if driver failed on this call, what is the behavior? How long the CSINode object will be kept with the past allocatable value? Or the failure to update means allocatable should be set to `0`?

Reply:
It should remain to the last reported value and retried IMO. This is no worse than current behaviour.